### PR TITLE
Make public/protected/private copy the method instead of creating a method entry using super

### PR DESCRIPTION
**Pull request by eregon**
Fixes:
* https://bugs.ruby-lang.org/issues/18729
* https://bugs.ruby-lang.org/issues/18435
* https://bugs.ruby-lang.org/issues/18751

And simplifies quite a lot of code by removing ZSUPER methods entirely.
matz approved to remove ZSUPER methods in https://bugs.ruby-lang.org/issues/18729#note-29 (will also be in the dev meeting log when it's published)

I think this nicely solves the design questions of @XrXr in  https://github.com/ruby/ruby/pull/5356#issuecomment-1005298809
In short, Method/UnboundMethod now always point to a given method definition which never changes (was the case before except for ZSUPER methods), owner is always the module which contains that method in its method table, and `Kernel#method`/`Module#instance_method` is more intuitive and returns the first method it finds (instead of looking above for ZSUPER methods in an attempt to hide them).

### CODE REVIEW

**Review by XrXr — COMMENTED** [review]
The new semantics does look nice to me. Since we're making a behavior change to [redefinition/removal scenarios](https://github.com/ruby/ruby/pull/5356#issuecomment-1005298809) to fix some bugs, assigning a new useful and unsurprising behavior makes sense.

With a quick read through, the patch looks straight-forward to me.

**Comment by matz** [comment]
Just for your record, I agree with #6242, not this one. I just gave up the illusion of only visibility is affected (without copies, without forwarding method entries).

**Comment by eregon** [comment]
OK, so I close this then, and I hope to merge #6242.
